### PR TITLE
[ICDS Dashboard] Data pull of children going school

### PR DESCRIPTION
**Pull request by Rohit25negi**
Urgent data pull for STQC.
Ticket: https://app.asana.com/0/1177124110930377/1180103810179979/f
Brief: this data pull contains the information of children going primary/private school for Jan/Feb/March

### CODE REVIEW

File: custom/icds_reports/management/commands/pull_Data_school.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+            for age_group in ('48', '60', '72'):
+
+                count_primary_school_going = 0
+                count_private_school_going = 0
+                count_joined_primary_school_during_month = 0
+                count_joined_private_school_during_month = 0
+                count_stopped_private_school_during_month = 0
+
+                person_case_ids = [item[0] for item in _run_custom_sql_script(query.format(age_group=age_group,month=month))]
+                child_person_cases = case_accessor.get_cases(person_case_ids)
```
Comment (mkangia) [inline custom/icds_reports/management/commands/pull_Data_school.py:41]: Do you want to consider `iter_cases` instead? I think this would be a huge number of cases

File: custom/icds_reports/management/commands/pull_Data_school.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+from django.core.management.base import BaseCommand
+from corehq.form_processor.interfaces.dbaccessors import CaseAccessors
+from django.db import connections, transaction
+from dateutil.relativedelta import relativedelta
+from corehq.sql_db.connections import get_icds_ucr_citus_db_alias
+
+
+def _run_custom_sql_script(command):
+    db_alias = get_icds_ucr_citus_db_alias()
```
Comment (mkangia) [inline custom/icds_reports/management/commands/pull_Data_school.py:11]: you can also just fetch it once globally.

File: custom/icds_reports/management/commands/pull_Data_school.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+                person_case_ids = [item[0] for item in _run_custom_sql_script(query.format(age_group=age_group,month=month))]
+                child_person_cases = case_accessor.get_cases(person_case_ids)
+
+                for case in child_person_cases:
+                    primary_school_admit = case.get_case_property('primary_school_admit')
+                    date_primary_admit = case.get_case_property('date_primary_admit')
+                    date_last_private_admit = case.get_case_property('date_last_private_admit')
+                    date_return_private = case.get_case_property('date_return_private')
+                    next_month = month + relativedelta(months=1)
+                    if primary_school_admit == 'yes' and date_primary_admit < next_month:
```
Comment (mkangia) [inline custom/icds_reports/management/commands/pull_Data_school.py:49]: `date_primary_admit < next_month`
are you comparing string with date?

Also, we don't need to worry about cases where these case properties don't have any value?

**Review by mkangia — COMMENTED** [review]
Can you add a comment to when each case property is set? Its hard to confirm the logic without knowing the purpose of these case properties.

File: custom/icds_reports/management/commands/pull_Data_school.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        for month in months:
+            for age_group in ('48', '60', '72'):
+
+                count_primary_school_going = 0
+                count_private_school_going = 0
+                count_joined_primary_school_during_month = 0
+                count_joined_private_school_during_month = 0
+                count_stopped_private_school_during_month = 0
+
+                person_case_ids = [item[0] for item in _run_custom_sql_script(query.format(age_group=age_group,month=month))]
```
[the remark this answers is not in the document]
Comment (ShubhamBansal1997) [reply custom/icds_reports/management/commands/pull_Data_school.py:40]: don't we need to change the month to string before passing to formatting

**Review by ShubhamBansal1997 — COMMENTED** [review]
A minor question otherwise code looks good. and could we use chunks like the entries gonna be too much

File: custom/icds_reports/management/commands/pull_Data_school.py (hunk 3 above)
Comment (Rohit25negi) [reply custom/icds_reports/management/commands/pull_Data_school.py:49]: hey, date_primary_admit case property contain the date.

yeah we are not worrying about those properties have any other values

Comment (Rohit25negi) [reply custom/icds_reports/management/commands/pull_Data_school.py:49]: you can check the tech spec of the datapull written added in description

File: custom/icds_reports/management/commands/pull_Data_school.py (hunk 2 above)
Comment (Rohit25negi) [reply custom/icds_reports/management/commands/pull_Data_school.py:11]: sure

File: custom/icds_reports/management/commands/pull_Data_school.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+                for case in child_person_cases:
+                    primary_school_admit = case.get_case_property('primary_school_admit')
+                    date_primary_admit = case.get_case_property('date_primary_admit')
+                    date_last_private_admit = case.get_case_property('date_last_private_admit')
+                    date_return_private = case.get_case_property('date_return_private')
+                    next_month = month + relativedelta(months=1)
+                    if primary_school_admit == 'yes' and date_primary_admit < next_month:
+                        count_primary_school_going += 1
+
+                    if date_last_private_admit < next_month and (date_return_private is None or date_return_private >= next_month or date_return_private == ''):
```
Comment (mkangia) [inline custom/icds_reports/management/commands/pull_Data_school.py:52]: nit: `not date_return_private` should be able to cover all null cases?

File: custom/icds_reports/management/commands/pull_Data_school.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+                    if primary_school_admit == 'yes' and date_primary_admit >= month and date_primary_admit < next_month:
+                        count_joined_primary_school_during_month += 1
+
+                    if date_last_private_admit >= month and date_last_private_admit < next_month:
+                        count_joined_private_school_during_month += 1
+
+                    if date_return_private >= month and date_return_private < next_month:
+                        count_stopped_private_school_during_month += 1
+
+                count_children_pse_in_awc = len(child_person_cases) - count_primary_school_going - count_private_school_going
```
Comment (mkangia) [inline custom/icds_reports/management/commands/pull_Data_school.py:64]: this does not match what is in spec excel sheet.

File: custom/icds_reports/management/commands/pull_Data_school.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+                        count_private_school_going += 1
+
+                    if primary_school_admit == 'yes' and date_primary_admit >= month and date_primary_admit < next_month:
+                        count_joined_primary_school_during_month += 1
+
+                    if date_last_private_admit >= month and date_last_private_admit < next_month:
+                        count_joined_private_school_during_month += 1
+
+                    if date_return_private >= month and date_return_private < next_month:
+                        count_stopped_private_school_during_month += 1
```
Comment (mkangia) [inline custom/icds_reports/management/commands/pull_Data_school.py:62]: nit: found this hard to follow so rewrote this to
```
if primary_school_admit == 'yes' and date_primary_admit < next_month:
    count_primary_school_going += 1
    if date_primary_admit >= month:
        count_joined_primary_school_during_month += 1

if date_last_private_admit < next_month:
    if date_return_private >= next_month or date_return_private is None or date_return_private == '':
        count_private_school_going += 1
    if date_last_private_admit >= month:
        count_joined_private_school_during_month += 1

if month <= date_return_private < next_month:
    count_stopped_private_school_during_month += 1
```

File: custom/icds_reports/management/commands/pull_Data_school.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Rohit25negi) [reply custom/icds_reports/management/commands/pull_Data_school.py:38]: no we dont. python handles it

File: custom/icds_reports/management/commands/pull_Data_school.py (hunk 6 above)
Comment (Rohit25negi) [reply custom/icds_reports/management/commands/pull_Data_school.py]: hey, do you want to offline. this actually matches that only.
PSE eligibles- row2(Public School Going Children) -row3(Private School Going Children)

Comment (mkangia) [reply custom/icds_reports/management/commands/pull_Data_school.py]: offlined

File: custom/icds_reports/management/commands/pull_Data_school.py [hunk 8]
```diff
@@ -36,32 +34,33 @@ def handle(self, **options):
...
-                for case in child_person_cases:
+                for case in case_accessor.iter_cases(person_case_ids):
+                    total_count += 1
                     primary_school_admit = case.get_case_property('primary_school_admit')
                     date_primary_admit = case.get_case_property('date_primary_admit')
                     date_last_private_admit = case.get_case_property('date_last_private_admit')
                     date_return_private = case.get_case_property('date_return_private')
                     next_month = month + relativedelta(months=1)
-                    if primary_school_admit == 'yes' and date_primary_admit < next_month:
+                    if primary_school_admit == 'yes' and parser.parse(date_primary_admit) < next_month:
```
Comment (mkangia) [inline custom/icds_reports/management/commands/pull_Data_school.py:47]: how about parsing this upfront? I see we need this at multiple places.

**Review by mkangia — APPROVED** [review]
looks good.

**Comment by orangejenny** [comment]
Can this be closed?
